### ISSUE TITLE
Nested inserts fail when there is more than one level of nesting

### ISSUE THREAD

**Issue by lovasoa** [hasura/graphql-engine#9709]
### Version Information

Server Version: v2.27.0

### How to reproduce the issue?

Create a postgres database with three tables a, b and c. 
c has a foreign key to b, which itself has a foreign key to a.

```sql
CREATE TABLE a (
    id serial primary key
);

CREATE TABLE b (
    id serial primary key references a(id)
);

CREATE TABLE c (
    id serial primary key references b(id),
    x text
);
```

connect the database to hasura and add all the relationships.

Try to insert a value into a, b and c using a nested insert [as documented](https://hasura.io/docs/latest/mutations/postgres/insert/#pg-nested-inserts)   

```graphql
mutation MyMutation {
  insert_a_one(object: {b: {data: {c: {data: {x: "Hello world"}}}}})
  { id }
}
```

### What is the current behaviour?

Hasura fails with the following error

```json
{
  "errors": [
    {
      "extensions": {
        "code": "validation-failed",
        "path": "$.selectionSet.insert_a_one.args.object[0].b.data[0].c"
      },
      "message": "cannot insert object relationship \"c\" as \"id\" column values are already determined"
    }
  ]
}
```

### What is the expected behaviour?

Hasura should execute something equivalent to this sql

```sql
WITH a_insert AS (INSERT INTO a VALUES (DEFAULT) RETURNING id),
     b_insert AS (INSERT INTO b SELECT id FROM a_insert RETURNING id),
     c_insert AS (INSERT INTO c SELECT id, 'Hello World' FROM b_insert)
SELECT id FROM a_insert;
```

### Screenshots or Screencast

![image](https://github.com/hasura/graphql-engine/assets/552629/b8e5fd78-b5a6-4f88-b227-b0c2a704b215)

![image](https://github.com/hasura/graphql-engine/assets/552629/a0836c7e-ef02-4744-9a1e-a2bbf697d640)

### Any possible solutions/workarounds you're aware of?

I would be keen on getting a good way to work around the problem.

In the meantime, one can

#### Start from "the middle"

If there are exactly three levels of nesting, one can start from the table that is "in the middle"; in our example: b.

```graphql
mutation MyMutation {
  insert_b_one(object: {
    a: {data:{}},
    c: {data:{x: "Hello world"}}
  }) { id }
}
```

#### Add a new manual relationship

A less than ideal solution is to manually add a direct relationship from a to c, change its insert_order in hasura's yaml metadata file (`metadata/databases/test/tables/public_a.yaml `) like so:

```yaml
table:
  name: a
  schema: public
object_relationships:
  - name: b
    using:
      foreign_key_constraint_on:
        column: id
        table:
          name: b
          schema: public
  - name: c
    using:
      manual_configuration:
        column_mapping:
          id: id
        insertion_order: after_parent # <- added manually in the yaml
        remote_table:
          name: c
          schema: public
```

and then run

```graphql
mutation MyMutation {
  insert_a_one(object: 
    	{
        b: {data: {}},
        c: {data: {x: "Hello world"}}
      }
  )
  { id }
}
``` 

### Keywords

hasura insert deeply nested fields

**Comment by manasag** [comment]
This is possibly same as #8522. The current solution seems to be same as you have already mentioned, using a manual insertion order.

**Comment by lovasoa** [comment]
@adas98012 : What is the reason for closing this as completed ? Do you prefer to follow up on this on #8522 ?

**Comment by manasag** [comment]
This has been addressed in v2.30.0-beta.1. 

We now skip the check that relationship columns can be over-specified when the insertion order is 'after_parent', allowing the parent row to provide the key columns for a child relationship.

Commit: https://github.com/hasura/graphql-engine/commit/57690491af9669420307a69c6d26d6204016e14f
